OpenSim/Moco/MocoFrameOrientationConstraint.cpp

```diff
@@ -0,0 +1,127 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "MocoFrameOrientationConstraint.h"
+
+using namespace OpenSim;
+
+
+MocoFrameOrientationConstraintPair::MocoFrameOrientationConstraintPair() {
+    constructProperties();
+}
+
+MocoFrameOrientationConstraintPair::MocoFrameOrientationConstraintPair(
+        std::string frame1Path, std::string frame2Path, double minimum_angle, 
+        double maximum_angle) {
+    constructProperties();
+    set_frame1_path(frame1Path);
+    set_frame2_path(frame2Path);
+    set_minimum_angle(minimum_angle);
+    set_maximum_angle(maximum_angle);
+}
+
+void MocoFrameOrientationConstraintPair::constructProperties() {
+    constructProperty_frame1_path("");
+    constructProperty_frame2_path("");
+    constructProperty_minimum_angle(-SimTK::Infinity);
+    constructProperty_maximum_angle(SimTK::Infinity);
+}
+
+//=============================================================================
+//  MocoFrameOrientationConstraint
+//=============================================================================
+
+MocoFrameOrientationConstraint::MocoFrameOrientationConstraint() {
+    constructProperties();
+}
+void MocoFrameOrientationConstraint::initializeOnModelImpl(const Model& model, 
+        const MocoProblemInfo&) const {
+
+    int nFramePairs = getProperty_frame_pairs().size();
+    MocoConstraintInfo info;
+    std::vector<MocoBounds> bounds;
+
+    for (int i = 0; i < nFramePairs; ++i) {
+        const auto frame1_path = get_frame_pairs(i).get_frame1_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame1_path), Exception,
+            "Could not find frame '{}'.", frame1_path);
+        auto& frame1 = model.getComponent<Frame>(frame1_path);
+        const auto frame2_path = get_frame_pairs(i).get_frame2_path();
+        OPENSIM_THROW_IF(!model.hasComponent<Frame>(frame2_path), Exception,
+            "Could not find frame '{}'.", frame2_path);
+        auto& frame2 = model.getComponent<Frame>(frame2_path);
+        m_frame_pairs.emplace_back(&frame1, &frame2);
+
+        const double& minimum = get_frame_pairs(i).get_minimum_angle();
+        const double& maximum = get_frame_pairs(i).get_maximum_angle();
+        
+        OPENSIM_THROW_IF(minimum > maximum, Exception,
+            "Expected the minimum angle for this frame pair "
+            "to be less than or equal to the maximum angle, "
+            "but they are {} and {}, respectively.",
+            minimum, maximum);
+        bounds.emplace_back(minimum, maximum);
+    }
+    m_coordinateAxis = get_coordinate_axis();

```

Comment:
Since this is optional, you'll want to check if it's empty (as that will determine if we minimize the 3D orientation error). You can check that the property is empty like so `getProperty_coordinate_axis().empty()`. You can also update the new boolean vector to use when calculating the path constraint error.